db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/fk/Graph.java

```diff
@@ -0,0 +1,181 @@
+    int surrogateCounter;
+    String newConstraintName;
+
+    public Graph(Set<String> vertices, String newConstraintName) {
+        edgeNodes = new ArrayList<>(vertices.size());
+        vertexIndex = new HashMap<>(vertices.size());
+        int i = 0;
+        for (String vertex : vertices) {
+            edgeNodes.add(null);
+            vertexIndex.put(vertex, i++);
+        }
+        parents = new int[vertices.size()][];
+        surrogateCounter = 0;
+        this.newConstraintName = newConstraintName;
+        this.surrogates = new HashMap<>();
+    }
+
+    void addEdge(String from, String to, EdgeNode.Type type) throws StandardException {
+        addEdgeInternal(vertexIndex.get(from), vertexIndex.get(to), type);
+    }
+
+    private void addSurrogate(int fromIdx, int toIdx, EdgeNode.Type type) {
+        int surrogateIdx = vertexIndex.size();
+        String name = getName(toIdx) + "__SURROGATE__" + surrogateCounter++;
+        surrogates.put(surrogateIdx, toIdx);
+        vertexIndex.put(name, surrogateIdx);
+        EdgeNode edgeNode = new EdgeNode(surrogateIdx, type);
+        edgeNode.next = edgeNodes.get(fromIdx);
+        edgeNodes.set(fromIdx, edgeNode);
+        edgeNodes.add(null); // for the surrogate
+    }
+
+    void addEdgeInternal(int fromIdx, int toIdx, EdgeNode.Type type) throws StandardException {
+        Dfs dfs = new Dfs(this, newConstraintName);
+        dfs.run(toIdx);
+        List<Integer> p = dfs.getPath(toIdx, fromIdx);
+        if(p.size() >= 2 && p.get(0) == toIdx && p.get(p.size() - 1) == fromIdx) {
+            breakCycle(p, type, fromIdx, toIdx);
+        } else {
+            EdgeNode edgeNode = new EdgeNode(toIdx, type);
+            edgeNode.next = edgeNodes.get(fromIdx);
+            edgeNodes.set(fromIdx, edgeNode);
+        }
+    }
+
+    private void breakCycle(List<Integer> p, EdgeNode.Type type, int fromIdx, int toIdx) throws StandardException {
+        if(type != EdgeNode.Type.C) {
+            addSurrogate(fromIdx, toIdx, type);
+            return;
+        } else {
+            for(int i = 1; i < p.size(); i++) { // find first none-C in the cycle and break it up
+                EdgeNode.Type edgeType = getEdgeType(p.get(i-1), p.get(i));
+                if(edgeType != EdgeNode.Type.C) {
+                    addSurrogate(p.get(i-1), p.get(i), edgeType);
+                    // remove this edge
+                    removeEdge(p.get(i-1), p.get(i));
+                    // add the new edge again, but check again for cycles!
+                    addEdgeInternal(fromIdx, toIdx, type);
+                    return;
+                }
+            }
+        }
+        // cycle is unbreakable, bail out
+        StringBuilder sb = new StringBuilder();
+        sb.append("adding the constraint between ").append(
+                getName(p.get(0))).append(" and ").append(getName(p.get(p.size()-1))).append(" would cause the following illegal delete action cascade cycle");
```

Comment:
i think normal + should be performant enough for this part of the code

sb.append("adding the constraint between " + getName(p.get(0)) + " and " + getName(p.get(p.size()-1)) + " would cause the following illegal delete action cascade cycle");

Reply:
I always try to avoid String concatenation and use a `StringBuilder` instead, I see your point, the code is ugly, I will make it more readable by restructuring it.

It is not only for performance, but it is kind of a habit I think :)

Let me know what you think about the new version once I push it.

Reply:
👍